drivers/gpio/gpio_altera_pio.c

```diff
@@ -0,0 +1,381 @@
+ *
+ * Set the port to 0. It will only write if data direction is set to output.
+ *
+ * @param dev  Pointer to the device structure for the driver instance.
+ * @param mask Value indicating which pins will be modified.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_altera_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	const int port_pin_mask = cfg->common.port_pin_mask;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	/* Check if mask range within 32 */
+	if ((port_pin_mask & mask) == 0) {
+		return -EINVAL;
+	}
+
+	if (!gpio_pin_direction(dev, mask)) {
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_GPIO_ALTERA_PIO_OUTCLEAR
+	addr = reg_base + ALTERA_AVALON_PIO_CLEAR_BITS;
+	sys_write32(mask, addr);
+#else
+	addr = reg_base + ALTERA_AVALON_PIO_DATA_OFFSET;
+	sys_clear_bits(addr, mask);
+#endif
+
+	return 0;
+}
+
+/**
+ * @brief Initialise an instance of the driver.
+ *
+ * This function initialise the interrupt configuration for the driver.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @return 0 to indicate success.
+ */
+static int gpio_init(const struct device *dev)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+
+	/* Configure GPIO device */
+	cfg->cfg_func();
+
+	return 0;
+}
+
+/**
+ * @brief Configure pin interrupt.
+ *
+ * @param dev	Pointer to the device structure for the driver instance.
+ * @param pin	Pin number to configure. (0 to 31)
+ * @param mode	Public flags. (GPIO_INT_MODE_DISABLED, GPIO_INT_MODE_LEVEL and
+ *				GPIO_INT_MODE_EDGE)
+ * @param trig	Public flags. (Edge Failing, Edge Rising or any)
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int gpio_altera_pin_interrupt_configure(const struct device *dev,
+						gpio_pin_t pin,
+						enum gpio_int_mode mode,
+						enum gpio_int_trig trig)
+{
+	ARG_UNUSED(trig);
+
+	const struct gpio_altera_config *cfg = dev->config;
+	uintptr_t reg_base = cfg->reg_base;
+	const int port_pin_mask = cfg->common.port_pin_mask;
+	uint32_t addr;
+
+	/* Check if pin number is within range */
+	if ((port_pin_mask & BIT(pin)) == 0) {
+		return -EINVAL;
+	}
+
+	addr = reg_base + ALTERA_AVALON_PIO_IRQ_OFFSET;
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:
+		/* Disable interrupt of pin */
+		sys_clear_bits(addr, BIT(pin));
+		break;
+	case GPIO_INT_MODE_LEVEL:
+	case GPIO_INT_MODE_EDGE:
+		/* Enable interrupt of pin */
+		sys_set_bits(addr, BIT(pin));
+		irq_enable(cfg->irq_num);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Generic function to insert or remove a callback from a callback list.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param callback	A valid Application’s callback structure pointer.
+ * @param set		A boolean indicating insertion or removal of the callback.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
```

Comment:
Please remove doxygen that duplicates (and possibly conflicts with) public driver api doxygen.